### PR TITLE
Hub Connection Builder

### PR DESCRIPTION
**Pull request by BrennanConroy**

### CODE REVIEW

File: include/signalrclient/hub_connection_builder.h [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+#include "_exports.h"
+#include "hub_connection.h"
+#include <memory>
+#include "websocket_client.h"
+#include "http_client.h"
+
+namespace signalr
+{
+    class hub_connection_builder
```
Comment (BrennanConroy) [inline include/signalrclient/hub_connection_builder.h]: This is the main thing to look at for now. Let's decide how we want the builder to look.

File: include/signalrclient/hub_connection.h [hunk 2]
```diff
@@ -46,6 +50,12 @@ namespace signalr
 
         SIGNALRCLIENT_API void send(const std::string& method_name, const signalr::value& arguments = signalr::value(), std::function<void(std::exception_ptr)> callback = [](std::exception_ptr) {}) noexcept;
 
+    protected:
+        friend class hub_connection_builder;
+
+        explicit hub_connection(const std::string& url, trace_level trace_level = trace_level::all,
```
Comment (halter73) [inline include/signalrclient/hub_connection.h]: Any reason not to just make the ctor private for now? Do we expect users to write classes that derive from hub_connection? In that case they wouldn't be able to use the builder right?

**Comment by halter73** [comment]
Any particular reason this is still a draft?

File: src/signalrclient/connection_impl.cpp [hunk 3]
```diff
@@ -52,6 +54,36 @@ namespace signalr
         }
     }
 
+    connection_impl::connection_impl(const std::string& url, trace_level trace_level, const std::shared_ptr<log_writer>& log_writer,
+        std::shared_ptr<http_client> http_client, std::function<std::shared_ptr<websocket_client>()> websocket_factory)
+        : m_base_url(url), m_connection_state(connection_state::disconnected), m_logger(log_writer, trace_level), m_transport(nullptr),
+        m_message_received([](const std::string&) noexcept {}), m_disconnected([]() noexcept {})
+    {
+        if (http_client != nullptr)
```
Comment (jkotalik) [inline src/signalrclient/connection_impl.cpp]: So the default for logging will be off? This is for connections, so I think it's okay.

File: src/signalrclient/hub_connection_builder.cpp [hunk 4]
```diff
@@ -0,0 +1,87 @@
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
```
Comment (jkotalik) [inline src/signalrclient/hub_connection_builder.cpp]:
```suggestion
// Copyright (c) .NET Foundation. All rights reserved.
// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
```

File: src/signalrclient/hub_connection_builder.cpp [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    hub_connection_builder& hub_connection_builder::with_http_client(std::shared_ptr<http_client> http_client)
+    {
+        m_http_client = http_client;
+
+        return *this;
+    }
+
+    hub_connection hub_connection_builder::build()
+    {
+#ifndef USE_CPPRESTSDK
```
Comment (jkotalik) [inline src/signalrclient/hub_connection_builder.cpp]: Isn't this being checked twice between here and connection_impl?

File: src/signalrclient/transport_factory.cpp [hunk 6]
```diff
@@ -5,16 +5,21 @@
...
+
     std::shared_ptr<transport> transport_factory::create_transport(transport_type transport_type, const logger& logger,
         const signalr_client_config& signalr_client_config)
     {
         if (transport_type == signalr::transport_type::websockets)
         {
-            return websocket_transport::create(
-                [signalr_client_config](){ return std::make_shared<default_websocket_client>(signalr_client_config); },
+            return websocket_transport::create(m_websocket_factory,
+                //[signalr_client_config](){ return std::make_shared<default_websocket_client>(signalr_client_config); },
```
Comment (jkotalik) [inline src/signalrclient/transport_factory.cpp]: nit: cleanup.

File: src/signalrclient/connection_impl.cpp (hunk 3 above)
Comment (BrennanConroy) [reply src/signalrclient/connection_impl.cpp]: Not sure what you mean here

**Comment by BrennanConroy** [comment]
> Any particular reason this is still a draft?

A little cleanup needed, and I was expecting some bikeshedding for the builder API :D

**Pull request merged**